TwilioVerifySDK/TwilioVerify/Sources/Data/Storage.swift

```diff
@@ -133,19 +151,56 @@ private extension Storage {
       try removeValue(for: result.key)
     }
   }
-  
-  func previousClearStorageOnReinstallValue() -> Bool? {
-    guard let clearStorageOnReinstallValue = try? get(Constants.clearStorageOnReinstallKey) else {
-      return nil
-    }
-    return Bool(String(decoding: clearStorageOnReinstallValue, as: UTF8.self))
-  }
-  
-  func updateClearStorageOnReinstall(value: Bool) {
-    guard let clearStorageOnReinstallValue = value.description.data(using: .utf8) else {
+
+  func checkAccessGroupMigration(
+    for accessGroup: String?,
+    using keychain: KeychainProtocol
+  ) {
+    if isAppExtension {
+      Logger.shared.log(withLevel: .debug, message: "Migrations are not available for app extensions.")
       return
     }
-    try? secureStorage.save(clearStorageOnReinstallValue, withKey: Constants.clearStorageOnReinstallKey, withServiceName: Constants.service)
+
+    let migrationDirection: MigrationDirection
+
+    switch (storedAccessGroup != nil, accessGroup != nil) {
+      case (false, true): migrationDirection = .toAccessGroup
+      case (true, false): migrationDirection = .fromAccessGroup
+      default: return
+    }
+
+    Logger.shared.log(withLevel: .debug, message: "Migrating Factors in direction: \(migrationDirection)")
+
+    var lastAccessGroup: String?
+
+    switch migrationDirection {
+      case .toAccessGroup:
+        guard let accessGroup = accessGroup else { return }
+        let factors = getAllFactors(using: factorMapper, keychain: keychain)
+
+        do {
+          try updateFactors(factors, with: accessGroup, keychain: keychain)
+          storedAccessGroup = accessGroup
+        } catch {
+          Logger.shared.log(
```

Comment:
Should we cancel the process for next steps like migrating the user defaults values?

Reply:
The `UserDefaults` values must be migrated even if the migration fails, because the `UserDefaults.suite` is being created previous this migration happen, so if for some reason it reach the `clearOnReinstall` validation, and these values does no exist in this `UserDefaults.suite`, could clear all data in keychain.

These `UserDefaults.suite`does not fail to create with a non valid app group, so these are easy to miss without taking notice of the issue.